### PR TITLE
Move parts of delegate to shared partition

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Delegate.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.Serialization;
+
+namespace System
+{
+    public abstract partial class Delegate : ICloneable, ISerializable
+    {
+		internal Delegate()
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Delegate.cs]: Nit: this file appears to be using tabs in many places... can you change it to use spaces instead?

File: src/System.Private.CoreLib/shared/System/Delegate.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
+using System.Reflection;
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Delegate.cs]: Missing license header.

File: src/System.Private.CoreLib/shared/System/Delegate.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Delegate.cs]: Why did this ctor change from private to internal?

File: src/System.Private.CoreLib/shared/System/Delegate.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		{
+			if (a is null)
+				return b;
+
+			return a.CombineImpl(b);
+		}
+
+		public static Delegate Combine(params Delegate[] delegates)
+		{
+			if (delegates == null || delegates.Length == 0)
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Delegate.cs]: Nit: it'd be good to be consistent, at least within the same file or least overloads of the same method, about how we do null checks.  You changed the above to be `a is null`... can you change this one to be `delegates is null`, too?

File: src/System.Private.CoreLib/shared/System/Delegate.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+			{
+				newDelegate = source;
+				source = Remove(source, value);
+			}
+			while (newDelegate != source);
+
+			return newDelegate;
+		}
+
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Delegate.cs]: Why did you remove all of the comments in these operators?  They were added very recently, in https://github.com/dotnet/coreclr/pull/21765.

File: src/System.Private.CoreLib/shared/System/Delegate.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		public static bool operator ==(Delegate d1, Delegate d2)
+		{
+			if (d2 is null)
+				return d1 is null;
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Delegate.cs]: Have you validated that the workaround that was employed previously is no longer necessary for performance?  The `return (d1 is null) ? true : false;` was there due to https://github.com/dotnet/coreclr/issues/914, which is still an open issue.  Same goes for the changes you made to the `return` statement below.

File: src/System.Private.CoreLib/shared/System/Delegate.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+			return ReferenceEquals(d2, d1) || d2.Equals((object) d1);
+		}
+
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		public static bool operator !=(Delegate d1, Delegate d2)
+		{
+			return !(d1 == d2);
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Delegate.cs]: Did you validate that this doesn't impact perf?

File: src/System.Private.CoreLib/shared/System/Delegate.cs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+			}
+			while (newDelegate != source);
+
+			return newDelegate;
+		}
+
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		public static bool operator ==(Delegate d1, Delegate d2)
+		{
+			if (d2 is null)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Delegate.cs]: This is missing micro-optimizations and related comments from the original version.

File: src/System.Private.CoreLib/shared/System/Delegate.cs (hunk 6 above)
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Delegate.cs]: Same

File: src/System.Private.CoreLib/shared/System/Delegate.cs (hunk 1 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Delegate.cs]: I used the version from corert, it's private now

File: src/System.Private.CoreLib/shared/System/Delegate.cs (hunk 3 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Delegate.cs]: done

File: src/System.Private.CoreLib/shared/System/Delegate.cs (hunk 4 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Delegate.cs]: I worked with several copied and forgot to align this one, should be now identical to coreclr

File: src/System.Private.CoreLib/shared/System/Delegate.cs (hunk 7 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Delegate.cs]: fixed

File: src/System.Private.CoreLib/shared/System/Delegate.cs (hunk 6 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Delegate.cs]: Used coreclr version but it's missing AggressiveInlining so not sure if it's intentional

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Delegate.cs]: @benaadams Was it intentional to mark `operator==` with AggressiveInlining, but not  `operator!=` ?

Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/Delegate.cs]: It was intentional to mark `operator==` with AggressiveInlining.

It would make sense to mirror it and mark `operator!=` with AggressiveInlining.

**Comment by marek-safar** [comment]
@jkotas anything else to be done here?

File: src/System.Private.CoreLib/shared/System/Delegate.cs (hunk 6 above)
Comment (danmosemsft) [reply src/System.Private.CoreLib/shared/System/Delegate.cs]: Why did such small methods not get inlined without the attribute?

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Delegate.cs]: This is not a small method after incorporating the other CR feedback.
